compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -360,13 +391,37 @@ impl<'a> Parser<'a> {
             }
             Err(err) => return Err(err),
         };
-        let ty = if self.eat(&token::Semi) {
-            TyKind::Array(elt_ty, self.parse_anon_const_expr()?)
-        } else {
-            TyKind::Slice(elt_ty)
-        };
-        self.expect(&token::CloseDelim(token::Bracket))?;
-        Ok(ty)
+        let size_const =
+            if self.eat(&token::Semi) { Some(self.parse_anon_const_expr()?) } else { None };
+
+        match self.expect(&token::CloseDelim(token::Bracket)) {
```

Comment:
I think that you could rewrite this as a single `match` on `self.token.kind` with four arms:

* `;`
* `,` if following thing could be const (pretty sure there's a method for that
* `]`
* _ case, where you call `expect_any(&[CloseDelim(Bracket), Semi])` purely for the error.

At that point I think you can entirely avoid the snapshotting (but can still do if you find cases where it is worth it).

Reply:
I like the idea of using a 4-way match here. I just implemented this and pushed.

I'm now using `struct_span_err`s to generate the errors. Good: The error message is now improved! When the recovery successful I now emit "expected `;`, found `,`" with a help, instead of "expected `;` or `]`, found `,`". Bad: the error message "expected ... found ..." will probably diverge from the one `expected_one_of_not_found` in time. I could call `expected_one_of_not_found`, but I'd have to add an `unreachable()!` to `Ok` case of the return value, which I'm not a fan of..

I think we could move the error string generation of `expected_one_of_not_found` and reuse it.

> , if following thing could be const (pretty sure there's a method for that

I looked for this function but couldn't find it. Assuming this function is similar to `is_certainly_not_a_block`, I'm also not a fan of this idea. The problem is these functions are usually overly conservative and break very easily, as shown in #82051.

It's not clear to me what the problem is with recovering as currently implemented. If it's simply confusing code or something like that, perhaps we can implement a helper function to encapsulate this pattern (save the parser state, restore parser state on error, continue on success).

Reply:
> I could call expected_one_of_not_found, but I'd have to add an unreachable()! to Ok case of the return value, which I'm not a fan of..

I think doing that would be reasonable to do.